### ISSUE TITLE
Provide TypeScript definition files

### ISSUE THREAD

**Issue by rcjsuen** [nodegit/nodegit#1189]
Since the JavaScript code is already generated, perhaps TypeScript definition files could also be generated?

Alternatively, I suppose they could be handwritten if it came down to it...

**Comment by johnhaley81** [comment]
We use [Flow](https://flowtype.org/) in GitKraken for our type system. I'd love to get some sort of typings up for NodeGit. A system that can send out both would be ideal.

We'd be able to maintain the flow stuff pretty well and confirm that it's valid since we use it daily in our environment but my only real concern is that nobody on the team uses TypeScript so it would be easier for an error to go under the radar for that type system.

**Comment by FlorianKoerner** [comment]
https://github.com/RabbitLabs/nodegit-typescript-definition

**Comment by dolanmiu** [comment]
I have made incredibly high quality typings for `nodegit`:

To install, simply do:

```bash
$ npm install @types/nodegit --save
```

Here is the pull request if interested:

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17102#issuecomment-308895183

It was hard, but I had fun 😄

**Comment by Stradivario** [comment]
@dolanmiu i can imagine ! Thank you :D

**Comment by Stradivario** [comment]
@dolanmiu 
![screenshot from 2018-03-21 12-49-01](https://user-images.githubusercontent.com/19847933/37705923-439341a2-2d06-11e8-9cbb-13b1789e9026.png)

I got the following error....

**Comment by dolanmiu** [comment]
@Stradivario I don't think this is an issue with the typings itself, can you try and see if installing another @types work?

**Comment by demurgos** [comment]
@Stradivario The npm client has many stability issues. Try to reinstall it, clear the cache, etc. You may also be interested in trying out _yarn_.

@dolanmiu Thanks for the typings. I am starting to look at them: it looks pretty good.

Ideally, the type definitions should be hosted here for documentation purposes. It offers a less ambiguous documentation and many tools started to use TS types to provide autocompletion and quick doc, even for JS projects. It would also help to keep the type definitions in sync with the code. (The types are currently at version `0.18` while the lib itself is at version `0.21`)

**Comment by Stradivario** [comment]
Yarn has more issues than npm and it is slower than npm on node 8.9.0 with 5.8.0 npm version my docker containers say that :) this is why i decided to revert back to npm with some docker caching layer also i am getting benefit from package-lock where without it build is with 10 seconds faster but what ever it is a different discussion.

I think i was having problems with DNS 8.8.8.8 google dns now i am happy user using nodegit typings thank you

**Comment by SlickNutter** [comment]
@dolanmiu `@types/nodegit` doesn't export `RepositoryInitOptions`

**Comment by dolanmiu** [comment]
I think it does:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/nodegit/repository.d.ts#L26-L34

**Comment by dolanmiu** [comment]
Right, I see

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32534

Here you go
